DOCS-8332 Merged reconnection strategy concept and reference (#1527)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/reconnection-strategy-about.adoc

include::_attributes.adoc[]
endif::[]
:keywords: anypoint, studio, reconnection strategies, reconnection strategy, retry policies, retry
:page-aliases: reconnection-strategy-reference.adoc

> When an operation in a Mule application fails to connect to an external server, the default behavior is for the operation to fail immediately. Upon failing, the operation returns a connectivity error.

Reviewer: Completely subjective:
```suggestion
When an operation in a Mule application fails to connect to an external server, the default behavior is for the operation to fail immediately and return a connectivity error.
```

> For example, if an operation in Anypoint Connector for FTP (FTP Connector) fails to connect to an FTP server, the default behavior is to throw an `FTP:CONNECTIVITY` error. The error name prepends a namespace that is specific to the connector, such as `FTP:CONNECTIVITY`, `HTTP:CONNECTIVITY`, or `DB:CONNECTIVITY` (for Anypoint Connector for Database (Database Connector).

Reviewer: Likewise subjective:
```suggestion
For example, if an operation in Anypoint Connector for FTP (FTP Connector) fails to connect to an FTP server, the operation throws an `FTP:CONNECTIVITY` error. The error name reflects the namespace that is specific to the connector, such as `FTP:CONNECTIVITY`, `HTTP:CONNECTIVITY`, or `DB:CONNECTIVITY` (for Anypoint Connector for Database (Database Connector).
```

> You can modify the default behavior by configuring a reconnection strategy on the operation.

Reviewer: ```suggestion
You can modify this default behavior by configuring a reconnection strategy for the operation.
```

> == Configure a Reconnection Strategy

Reviewer: ```suggestion
== Configure an Operation Reconnection Strategy
```

> You can configure a reconnection strategy by modifying the operation properties, or by modifying the global element configuration for the desired operation. For example, you can configure a reconnection strategy in an HTTP Request configuration:

Reviewer: ```suggestion
You can configure a reconnection strategy for an operation either by modifying the operation properties or by modifying the configuration of the global element for the operation. For example, you can configure a reconnection strategy in an HTTP Request configuration:
```

image::mruntime-reconnection-strategy.png[Configure a reconnection strategy]

> The following are the available reconnection strategies and their behavior:

Reviewer: ```suggestion
The following are the available reconnection strategies and their behaviors:
```

> * *None*

Reviewer: Why are these names in bold?
We typically reserve bold for UI labels.
Ditto all three.

+

> The default behavior, which immediately returns a connectivity error if the attempt to connect is unsuccessful.

Reviewer: These list-item descriptions are not complete sentences, so...no periods.
Added "is" for list-item parallelism (verb-based):
```suggestion
Is the default behavior, which immediately returns a connectivity error if the attempt to connect is unsuccessful
```

* *Standard* (`reconnect`)
+

> Sets a number of reconnection attempts and the interval at which to execute them before returning a connectivity error.

Reviewer: ```suggestion
Sets the number of reconnection attempts and the interval at which to execute them before returning a connectivity error
```

* *Forever* (`reconnect-forever`)
+

> Attempts to reconnect continually at a given interval.

Reviewer: ```suggestion
Attempts to reconnect continually at a given interval
```
Do we need to explain when or why this strategy would be chosen?

> == XML Examples

Reviewer: ?
```suggestion
== Example XML Reconnection Strategy Configurations
```
It probably seems like overkill, but I keep trying to remember to look at the wording of titles so that they are very precise and make sense even taken out of the context of their containers; this also helps with info searches, I think.

> The following examples show different types of reconnection strategies and their configuration.

Reviewer: ```suggestion
The following examples show different types of reconnection strategies and their configurations:
```

> * Example: Standard reconnection strategy

Reviewer: ```suggestion
* Standard reconnection strategy example
```

+

> This example defines a Standard reconnection strategy in an FTP connector configuration, setting a `count` of `5` reconnection attempts before failing, and setting `failsDeployment` to `true`, causing the application deployment to fail if the connection test fails. Because the `frequency` is not specified, the retry interval is every `2000` milliseconds (ms), which is the default value:

Reviewer: Hmm...I think we can introduce this using general words rather than syntax:
```suggestion
The following example illustrates a Standard reconnection strategy in an FTP connector configuration, setting the count of reconnection attempts before failing to `5` and setting `failsDeployment` to `true`, which causes the application deployment to fail if the connection test fails. Because a frequency is not specified, the retry interval is the default of every 2000 ms:
```
```

  </ftp:listener>
</flow>
----

> * Example: Forever reconnection strategy

Reviewer: ```suggestion
* Forever reconnection strategy example
```

+

> This example sets a Forever reconnection strategy in an FTP connector. The connector attempts to reconnect every `4000` ms for as long as the application runs:

Reviewer: ```suggestion
The following example sets a Forever reconnection strategy in an FTP connector configuration by which the connector attempts to reconnect every 4000 ms for as long as the application runs:
```

</ftp:connection>
----

> == Reconnection Strategy Reference

Reviewer: ```suggestion
== Reconnection Strategy Configuration Attribute Reference
```
OR
```suggestion
== Reconnection Strategy Configuration Attributes
```

Following are the different configuration attributes for each reconnection strategy you can configure by editing your Mule application's XML configuration file:

* Attributes of `<reconnection>`
+
[%header%autowidth.spread]
|===
|Name |Type |Required |Default |Description

> |`failsDeployment` |boolean |no |false | If set to `true`, causes the deployment to fail when the test connection fails

Reviewer: ```suggestion
|`failsDeployment` |boolean |no |false | If `true`, causes the deployment to fail when the test connection fails
```

|===
* Attributes of `<reconnect>`
+
[%header%autowidth.spread]
|===
|Name |Type |Required |Default |Description

> |`blocking` |boolean |no |true |If set to `false`, the reconnection strategy runs in a separate, non-blocking thread

Reviewer: ```suggestion
|`blocking` |boolean |no |true |If `false`, the reconnection strategy runs in a separate, nonblocking thread
```

|`frequency` |long |no |2000 |How often (in ms) to reconnect
|`count` |integer |no |2 |How many reconnection attempts to make
|===
* Child Elements of `<reconnect>`
+
[%header%autowidth.spread]
|===
|Name |Cardinality |Description

> |`abstract-reconnect-notifier` |0..1 |A placeholder for a reconnection notifier element. The RetryNotifier interface is a callback that allows actions to be performed after each reconnection attempt, for example, firing server notification events on success or failure.

Reviewer: I don't know if this formatting works in a table, but...:
```suggestion
|`abstract-reconnect-notifier` |0..1 |A placeholder for a reconnection notifier element +
The RetryNotifier interface is a callback that allows actions to be performed after each reconnection attempt, for example, firing server notification events on success or failure.
```

|===
* Attributes of `<reconnect-forever>`
+
[%header%autowidth.spread]
|===
|Name |Type |Required |Default |Description

> |`blocking` |boolean |no |true |If false, the reconnection strategy runs in a separate, non-blocking thread.

Reviewer: ```suggestion
|`blocking` |boolean |no |true |If `false`, the reconnection strategy runs in a separate, nonblocking thread.
```

> |`frequency` |long |no |2000 |How often (in ms) to reconnect.

Reviewer: Just to make it a parallel sentence....
```suggestion
|`frequency` |long |no |2000 |Specifies how often (in ms) to reconnect.
```
